utils/docker/Dockerfile.centos.7

```diff
@@ -141,6 +141,16 @@ RUN ln -s utils/sl site_scons
 # ease-of-use for users, however in CI everything is turned off and then
 # selectively enabled.  Turning off any step disables all future steps.
 ARG DAOS_DEPS_BUILD=yes
+
+# Now do an update to ensure software is up to date for the deps build.  If the
+# src hasn't changed then this won't do anything, but if it has then we want to
+# ensure that latest dependencies are used.
+USER root:root
+RUN [ "$DAOS_DEPS_BUILD" != "yes" ] || \
+    { yum -y upgrade --exclude=spdk,spdk-devel,dpdk-devel,dpdk,mercury-devel,mercury && \
+    yum clean all; }
```

Comment:
Just confirming intention here... if an O/S supplied `-devel` for one of the dependencies that we build was updated since the last time we built the dependency (which is now stored in a docker layer, and will continue to be used for future docker builds), it's OK that we are not rebuilding the dependency here to get the updated `-devel` contents?

I would think we would most certainly want to rebuild one of our dependencies with an O/S supplied (i.e. `-devel`) dependency that bumped it's major version (indicating incompatible ABI).  I'm on the fence for < $major_update updates.  I could imagine that as long as it's not a major version bump, API (`*.h` files) should not be changing and the dynamic linker will still be supplying a binary compatible (newer) library at run-time.

The major version bump is concerning though, and what happen here when that happens?

Reply:
Yes, I think that's OK.  Distros do not break API within minor versions of the distro and "yum upgrade" will not update major versions.

There is the potential for our dependencies to be built with one version of a system package and DAOS to be built with a newer version, but that's true of almost anything built from source, and upstream updates could already mean it's tested with a newer version again.  We'll be at most a week out of date however, and as I said distros don't break compatibility in point releases.  Any PR that modifies the dependences (via build.config or new build instructions) will build the dependencies and DAOS with the same version.

Reply:
> Yes, I think that's OK. Distros do not break API within minor versions of the distro

That's not the only breakage scenario.  Distros do update minor versions within (or at least across) point releases and we could bump up the version of one of our components (libfabric as a good example of something we sit on the bleeding edge of) that could want ABI that has been **added** (so does not require a major version bump, just a minor) to a library being supplied by the O/S.  Maybe that's a small-to-non-existent case though.

> and "yum upgrade" will not update major versions.

That's not true at all.  `yum upgrade` will happily upgrade any package version to any other higher package version of the same name (or specified `Obsoletes:`) without any regard to what's in the RPM itself, major library versions included.  It is up to packagers to ensure that they are not causing breakage by packaging something that will break other things that depend on what they are packaging.

Now, in terms of libraries, I do think this is one place that SUSE (and Debian/Ubuntu) get right where RH does not.  SUSE packages their libraries with the major version in the name (i.e. `libfoo1`, `libfoo2`, etc.) to give packagers more freedom to update library major versions without removing previous major versions and causing such breakage.

RH handles this in a more limited way by adding `-compat` packages when they need more than 1 library major version to exist.

Reply:
Updating distro packages that include bugfixes or add features is not going to be a problem.

When I said "major versions" I was referring to the distro itself, not about packages.  That said, if one of our dependences depends on "libfoo" version 1 then calling yum upgrade isn't going to remove version 1 and leave only version 2, if that were true it would be almost impossible to every update after building things from source.

Distros just don't break the ABI in minor releases, and certainly not if you're only stepping forwards not backwards.